File: March2022/Java/CountAllValidPickupAndDeliveryOptions.java

// @saorav21994
// For arranging all pickup we will have n! options as relative arrangement among them does not matter
// For arranging all delivery we will have 1*3*5*3...*(2n-1) options as D[i] can come at any place but only after P[i]
// TC : O(n)
// SC : O(1)

// In C++, bottom up approach solution is given


class Solution {
    
    public int countOrders(int n) {
        
        long res = 1l;
        final int MOD = 1000000007;
        
        for (int i = 1; i <= n; i++) {
            res *= i;           // n! for pickup arrangement
            res %= MOD;
            res *= (2*i-1);     // 1*3*5*...(2n-1) for delivery
            res %= MOD;
        }
        return (int)res;
    }
}

File: March2022/C++/CountAllValidPickupAndDeliveryOptions.cpp

// @saorav21994
// TC : O(n^2)
// SC : O(n^2)

// Bottom up approach, DP -> recurrence relation => dp[i][j] = (unpicked order) * countOrder(unpicked-1, undelivered) + (orderPicked but undelivered) * countOrder(unpicked, undelivered-1)


class Solution {
public:
    int countOrders(int n) {
        int MOD = 1000000007;
        
        int i = 0;  // unpicked items
        int j = 0;  // undelivered items
        
        long dp[n+1][n+1];
    
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= n; j++) {
                dp[i][j] = 0;
            }
        }
        
        for (i = 0; i <= n; i++) {
            for (j = i; j <= n; j++) {
                if (i == 0 && j == 0) {
                    dp[i][j] = 1;
                    continue;
                }
                
                if (i > 0) {
                    dp[i][j] += i * dp[i-1][j];
                }
                dp[i][j] %= MOD;
                
                if (j > i) {
                    dp[i][j] += (j-i) * dp[i][j-1];
                }
                dp[i][j] %= MOD;
                
            }
        }
        return (int)dp[n][n]; 
    }
};
